### ISSUE TITLE
Firefox 47 doesn't work with Galen Framework 2.2.5

### ISSUE THREAD

**Issue by tansin** [galenframework/galen#424]
After automated update of Firefox to 47.0 Galen tests don't run anymore.
The problem seems to be a known issue with selenium not working with Firefox 47 in conjunction with extension signing feature. See: (https://github.com/seleniumhq/selenium/issues/2110).

This issue contains the advise to use Marionette as the only FirefoxDriver which will be supported in future (https://developer.mozilla.org/en-US/docs/Mozilla/QA/Marionette/WebDriver).

There are many more recent issues relating to the same problem with selenium. 

**OS:** OS X El Capitan 10.11.5
**java version:** 1.8.0_74
**galen version:** 2.2.5
**used browser and version:** Firefox 47.0

**steps leading to the issue:**
Start test using a firefox driver, e.g.:
`createDriver("http://localhost:3000", "1280x800", "firefox");`

**error log:**
Fatal error: Command failed: galen test tests/galen.test.js --htmlreport galen-report 
org.openqa.selenium.firefox.NotConnectedException: Unable to connect to host localhost on port 7055 after 45000 ms. Firefox console output:
8-4474-a285-3208198ce6fd}","syncGUID":"0wvGOi5CNeUM","location":"app-global","version":"47.0","type":"theme","internalName":"classic/1.0","updateURL":null,"updateKey":null,"optionsURL":null,"optionsType":null,"aboutURL":null,"icons":{"32":"icon.png","48":"icon.png"},"iconURL":null,"icon64URL":null,"defaultLocale":{"name":"Default","description":"The default theme.","creator":"Mozilla","homepageURL":null,"contributors":["Mozilla Contributors"]},"visible":true,"active":true,"userDisabled":false,"appDisabled":false,"descriptor":"/Applications/Firefox.app/Contents/Resources/browser/extensions/{972ce4c6-7e08-4474-a285-3208198ce6fd}.xpi","installDate":1465092614000,"updateDate":1465092614000,"applyBackgroundUpdates":1,"skinnable":true,"size":7134,"sourceURI":null,"releaseNotesURI":null,"softDisabled":false,"foreignInstall":false,"hasBinaryComponents":false,"strictCompatibility":true,"locales":[],"targetApplications":[{"id":"{ec8030f7-c20a-464f-9b0e-13a3a9e97384}","minVersion":"47.0","maxVersion":"47.0"}],"targetPlatforms":[],"seen":true}
1465991649765	addons.xpi	DEBUG	getModTime: Recursive scan of {972ce4c6-7e08-4474-a285-3208198ce6fd}
1465991649766	DeferredSave.extensions.json	DEBUG	Save changes
1465991649766	addons.xpi	DEBUG	Updating database with changes to installed add-ons
1465991649766	addons.xpi-utils	DEBUG	Updating add-on states
1465991649767	addons.xpi-utils	DEBUG	Writing add-ons list
1465991649768	addons.xpi	DEBUG	Registering manifest for /Applications/Firefox.app/Contents/Resources/browser/features/e10srollout@mozilla.org.xpi
1465991649769	addons.xpi	DEBUG	Calling bootstrap method startup on e10srollout@mozilla.org version 1.0
1465991649769	addons.xpi	DEBUG	Registering manifest for /Applications/Firefox.app/Contents/Resources/browser/features/firefox@getpocket.com.xpi
1465991649769	addons.xpi	DEBUG	Calling bootstrap method startup on firefox@getpocket.com version 1.0.2
1465991649770	addons.xpi	DEBUG	Registering manifest for /Applications/Firefox.app/Contents/Resources/browser/features/loop@mozilla.org.xpi
1465991649771	addons.xpi	DEBUG	Calling bootstrap method startup on loop@mozilla.org version 1.3.2
1465991649792	addons.manager	DEBUG	Registering shutdown blocker for XPIProvider
1465991649792	addons.manager	DEBUG	Provider finished startup: XPIProvider
1465991649792	addons.manager	DEBUG	Starting provider: LightweightThemeManager
1465991649792	addons.manager	DEBUG	Registering shutdown blocker for LightweightThemeManager
1465991649792	addons.manager	DEBUG	Provider finished startup: LightweightThemeManager
1465991649793	addons.manager	DEBUG	Starting provider: GMPProvider
1465991649799	addons.manager	DEBUG	Registering shutdown blocker for GMPProvider
1465991649799	addons.manager	DEBUG	Provider finished startup: GMPProvider
1465991649799	addons.manager	DEBUG	Starting provider: PluginProvider
1465991649799	addons.manager	DEBUG	Registering shutdown blocker for PluginProvider
1465991649799	addons.manager	DEBUG	Provider finished startup: PluginProvider
1465991649800	addons.manager	DEBUG	Completed startup sequence
1465991650128	addons.manager	DEBUG	Starting provider: <unnamed-provider>
1465991650128	addons.manager	DEBUG	Registering shutdown blocker for <unnamed-provider>
1465991650129	addons.manager	DEBUG	Provider finished startup: <unnamed-provider>
1465991650135	DeferredSave.extensions.json	DEBUG	Starting write
1465991650279	addons.repository	DEBUG	No addons.json found.
1465991650280	DeferredSave.addons.json	DEBUG	Save changes
1465991650284	DeferredSave.addons.json	DEBUG	Starting timer
1465991650326	addons.manager	DEBUG	Starting provider: PreviousExperimentProvider
1465991650327	addons.manager	DEBUG	Registering shutdown blocker for PreviousExperimentProvider
1465991650327	addons.manager	DEBUG	Provider finished startup: PreviousExperimentProvider
1465991650335	DeferredSave.extensions.json	DEBUG	Write succeeded
1465991650336	addons.xpi-utils	DEBUG	XPI Database saved, setting schema version preference to 17
1465991650337	DeferredSave.addons.json	DEBUG	Starting write
1465991650372	DeferredSave.addons.json	DEBUG	Write succeeded

	at org.openqa.selenium.firefox.internal.NewProfileExtensionConnection.start(NewProfileExtensionConnection.java:113)
	at org.openqa.selenium.firefox.FirefoxDriver.startClient(FirefoxDriver.java:271)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:119)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:216)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:211)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:128)
	at com.galenframework.browser.SeleniumBrowserFactory.getDriver(SeleniumBrowserFactory.java:95)
	at com.galenframework.browser.SeleniumBrowserFactory.createLocalBrowser(SeleniumBrowserFactory.java:89)
	at com.galenframework.browser.SeleniumBrowserFactory.openBrowser(SeleniumBrowserFactory.java:65)
	at com.galenframework.utils.GalenUtils.createDriver(GalenUtils.java:262)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.mozilla.javascript.MemberBox.invoke(MemberBox.java:126)
	at org.mozilla.javascript.NativeJavaMethod.call(NativeJavaMethod.java:225)
	at org.mozilla.javascript.optimizer.OptRuntime.callN(OptRuntime.java:52)
	at org.mozilla.javascript.gen._cmd__18._c_createDriver_2(<cmd>:37)
	at org.mozilla.javascript.gen._cmd__18.call(<cmd>)
	at org.mozilla.javascript.optimizer.OptRuntime.callName(OptRuntime.java:63)
	at org.mozilla.javascript.gen._Users_user1_tests____tests_gl_js_21._c_getSingleDeviceTarget_18(/Users/user1/tests/../tests/gl.js:346)
	at org.mozilla.javascript.gen._Users_user1_tests____tests_gl_js_21.call(/Users/user1/tests/../tests/gl.js)
	at org.mozilla.javascript.optimizer.OptRuntime.callName(OptRuntime.java:63)
	at org.mozilla.javascript.gen._Users_user1_tests____tests_gl_js_21._c_openPage_17(/Users/user1/tests/../tests/gl.js:319)
	at org.mozilla.javascript.gen._Users_user1_tests____tests_gl_js_21.call(/Users/user1/tests/../tests/gl.js)
	at org.mozilla.javascript.optimizer.OptRuntime.call2(OptRuntime.java:42)
	at org.mozilla.javascript.gen._Users_user1_tests_galen_test_js_20._c_anonymous_3(/Users/user1/tests/galen.test.js:29)
	at org.mozilla.javascript.gen._Users_user1_tests_galen_test_js_20.call(/Users/user1/tests/galen.test.js)
	at org.mozilla.javascript.ScriptRuntime.applyOrCall(ScriptRuntime.java:2429)
	at org.mozilla.javascript.BaseFunction.execIdCall(BaseFunction.java:269)
	at org.mozilla.javascript.IdFunctionObject.call(IdFunctionObject.java:97)
	at org.mozilla.javascript.optimizer.OptRuntime.callN(OptRuntime.java:52)
	at org.mozilla.javascript.gen._cmd__17._c_invokeFunc_2(<cmd>:35)
	at org.mozilla.javascript.gen._cmd__17.call(<cmd>)
	at org.mozilla.javascript.optimizer.OptRuntime.callName(OptRuntime.java:63)
	at org.mozilla.javascript.gen._cmd__17._c_anonymous_22(<cmd>:280)
	at org.mozilla.javascript.gen._cmd__17.call(<cmd>)
	at org.mozilla.javascript.ContextFactory.doTopCall(ContextFactory.java:394)
	at org.mozilla.javascript.ScriptRuntime.doTopCall(ScriptRuntime.java:3090)
	at org.mozilla.javascript.gen._cmd__17.call(<cmd>)
	at org.mozilla.javascript.JavaAdapter.doCall(JavaAdapter.java:606)
	at org.mozilla.javascript.JavaAdapter.access$000(JavaAdapter.java:15)
	at org.mozilla.javascript.JavaAdapter$1.run(JavaAdapter.java:586)
	at org.mozilla.javascript.Context.call(Context.java:489)
	at org.mozilla.javascript.ContextFactory.call(ContextFactory.java:504)
	at org.mozilla.javascript.JavaAdapter.callMethod(JavaAdapter.java:583)
	at adapter3.execute(<adapter>)
	at com.galenframework.TestRunnable.runTest(TestRunnable.java:75)
	at com.galenframework.TestRunnable.run(TestRunnable.java:101)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

**Comment by hypery2k** [comment]
For new Firefox we need to support the MarionetteWebDriver. But I think we need to first somehow detect if new firefox is working and then use this.

**Comment by ishubin** [comment]
Thanks for reporting! I will try to see if it can be easily fixed

**Comment by marthijsBerfelo** [comment]
I've recently implemented MarionetteDriver in a JAVA framework and a Fitnesse/Selenium project. MarionetteDriver's latest release is still a beta version and is not releasing ports correclty, so is not yet ready enough to use when running multiple tests without completely quitting the test-application. 

Since Galen uses selenium the problem is related to a bug in FF47 that should be resolved with the next release see (https://github.com/SeleniumHQ/selenium/issues/2110).

For now the only real solution I have found is to downgrade to FF46.

**Comment by ishubin** [comment]
@marthijsBerfelo I agree. Lets wait until it gets stable

**Comment by aspirisen** [comment]
Hi, i am a newcomer and i just wanted to see what is galen and I've faced this problem too. It took a lot of time to get what was the problem. It seems that everyone who will download galen will fall into this issue and they will not be able to try this tool. I think it should be warned somewhere on the site and in readme file or use another browser.

**Comment by ishubin** [comment]
@aspirisen Thanks for the good idea! I have just updated the website with red warning on every page and also added a dedicated page which explains how to configure other browsers: http://galenframework.com/docs/browsers/

What do you think about that?

**Comment by aspirisen** [comment]
@ishubin great, now it can help for new people who just start using Galen. Thanks

**Comment by marthijsBerfelo** [comment]
@ishubin @hypery2k The issue with Firefox and Selenium has been fixed and new versions have been released. When using selenium v2.53.1 and Firefox 47.0.1, the browser can be started without immediately crashing.

I would like to help out and change the version

**Comment by hypery2k** [comment]
@ishubin Maybe we could make a bug fixing release for 2.2 with updated Selenium?

**Comment by ishubin** [comment]
@hypery2k Sure! I will make a release today

**Comment by hypery2k** [comment]
Cool, thanks

**Comment by ishubin** [comment]
I have just released an update in 2.2.6 and tested it with Firefox 47.0.1. For some reason it didn't work with Firefox 47.0.0 but was fine with 47.0.1

**Comment by marthijsBerfelo** [comment]
Hi Ivan,

47.0.0 has a problem with xpi signing in the selenium plugin. This was fixed in 47.0.1 and selenium 2.53.1 was released for an new Firefox driver



Met vriendelijke groet,

Marthijs Berfelo
Technisch testspecialist
Polteq

06 3190 3677
marthijs.berfelo@polteq.com


Verzonden vanaf mijn mobiele telefoon.


-------- Oorspronkelijk bericht --------
Van: Ivan Shubin <notifications@github.com>
Datum: 08-07-2016 22:04 (GMT+01:00)
Aan: galenframework/galen <galen@noreply.github.com>
Cc: Marthijs Berfelo <marthijs.berfelo@polteq.com>, Mention <mention@noreply.github.com>
Onderwerp: Re: [galenframework/galen] Firefox 47 doesn't work with Galen Framework 2.2.5 (#424)


I have just released an update in 2.2.6 and tested it with Firefox 47.0.1. For some reason it didn't work with Firefox 47.0.0 but was fine with 47.0.1

-
